src/cljc/schema/spec/collection.cljc

```diff
@@ -23,19 +23,18 @@
       (let [_ (macros/assert! (= 2 (count e)) "remaining can have only one schema.")
             c (spec/sub-checker (second e) params)]
         #?(:clj (fn [^java.util.List res x]
-                  (doseq [i x]
-                    (.add res (c i)))
+                  (reduce (fn [res i] (doto res (.add (c i)))) res x)
```

Comment:
Why is this expected to be faster?  The `List` can't be used safely concurrently so I would be surprised if you could do better than the `doseq`.  Have you profiled and seen a speedup?

Reply:
`reduce` (and `run!`) is faster and more memory efficient because as an internal iterator it can be fully specialized to the collection and usually does not need to allocate at all. I have profiled it sometime but by now it is just a habit like using `(into [] (map ...) x)` instead of `(into [] (map ... x))`. Since `reduce` works on all seqs really the only reason to ever use `doseq` is if `:when`/`:while`/`:let` is super useful.

Reply:
The weird thing about `reduce` here is that it's threading through an extra argument that you don't care about because you're mutating the argument anyway.   I think it makes the code slightly less clear and I would be a bit surprised if there was a performance benefit, but I'd be happy to believe it if there were accompanying measurements.

Reply:
> The weird thing about reduce here is that it's threading through an extra argument that you don't care about because you're mutating the argument anyway.

Might be worth comparing to `(reduce #(.add res (c %2))) nil x)`.

Reply:
Oh, it would probably beat the current code handily since I don't see a type hint on the inner `res`.